src/main/kotlin/compiler/Coder.kt

```diff
@@ -99,4 +99,90 @@ class Coder(val ast: Node) {
         }
         return s
     }
+
+
+    object Optimizer {
+        var source = ArrayList<VMWord>()
+        var mem = ArrayList<VMWord>()
+        val jumpMap = HashMap<Int, Int>()
+        var pc = 0
+        var lastMatchSize = 0
+
+        fun postOptimize(withSource: ArrayList<VMWord>) {
+
+            // Find all jump destinations in source
+            source = withSource
+            source.forEach { word ->
+                word.address?.also { address ->
+                    jumpMap[address] = -1
+                }
+            }
+
+            mem.clear()
+            pc = 0
+            while (pc < source.size) {
+                // If we've reached a jump dest, record its new address
+                if (jumpMap.containsKey(pc)) jumpMap[pc] = mem.size
+
+
+                // NEGATE NEGATE => ()
+                consume(O_NEGATE, O_NEGATE)?.also { }
+
+                // SETVAR GETVAR => SETGETVAR
+                ?: consume(O_SETVAR, null, O_GETVAR, null) { args ->
+                    args[0].isInt() && args[1].isInt(args[0].intFromV)
+                }?.also { args ->
+                    code(O_SETGETVAR)
+                    value(args[0].value!!)
+                }
+
+                // O_VAL 0 O_CMP_xx => O_CMP_xxZ
+                ?: consume(O_VAL, null, O_CMP_EQ) { args -> args[0].isInt(0) }?.also { code(O_CMP_EQZ) }
+                ?: consume(O_VAL, null, O_CMP_GT) { args -> args[0].isInt(0) }?.also { code(O_CMP_GTZ) }
+                ?: consume(O_VAL, null, O_CMP_GE) { args -> args[0].isInt(0) }?.also { code(O_CMP_GEZ) }
+                ?: consume(O_VAL, null, O_CMP_LT) { args -> args[0].isInt(0) }?.also { code(O_CMP_LTZ) }
+                ?: consume(O_VAL, null, O_CMP_LE) { args -> args[0].isInt(0) }?.also { code(O_CMP_LEZ) }
+
+
+                // If nothing matched, copy and continue
+                ?: run {
+                    mem.add(source[pc++])
+                }
+            }
+
+            // Replace all jump dests
+            jumpMap.keys.forEach { old ->
+                mem.forEach { word ->
+                    if (word.address == old) word.address = jumpMap[old]
+                }
+            }
```

Comment:
Making this map and applying it at the end seems fine, but it occurs to me that maybe you could adjust the jump addresses as you traverse the source (and avoid this O(2n) remapping operation) by keeping track of how many words have been removed and subtracting that value from the jump addresses as you encounter them in your optimization loop.

Reply:
That is actually what I did in my first draft!  I didn't like how it looked, it felt messy, so I stared at it for a while until I realized I could just write them down as I reached them, and then fill them in at the end.  You're right though, the method you describe does indeed do the trick.

I see what you mean though about it being more efficient.  I need to ponder a bit about how big that impact could be; I might change back.  Thanks for bringing this up.